python/ray/dashboard/metrics_exporter/schema.py

```diff
@@ -30,41 +44,67 @@ def json(self):
 
     @classmethod
     def parse_obj(cls, obj):
+        # Validation.
         assert type(obj) == dict, ("It can only parse dict type object.")
-        required_args = cls.__slots__
-        given_args = obj.keys()
-
-        # Check if given_args have args that is not required.
-        for arg in given_args:
-            if arg not in required_args:
-                raise ValidationError(
-                    "Given argument has a key {}, which is not required "
-                    "by this schema: {}".format(arg, required_args))
-
-        # Check if given args have all required args.
-        if len(required_args) != len(given_args):
-            raise ValidationError("Given args: {} doesn't have all the "
-                                  "necessary args for this schema: {}".format(
-                                      given_args, required_args))
+        for field, schema in cls.__schema__.items():
+            required, default, arg_type = schema
+            if field not in obj:
+                if required:
+                    raise ValidationError(f"{field} is required, but doesn't "
+                                          "exist in a given object {obj}")
+                else:
+                    # Set default value if the field is optional
+                    obj[field] = default
 
         return cls(**obj)
 
 
+"""
+Request/Response
+"""
+
+
 class IngestRequest(BaseModel):
-    __slots__ = [
-        "cluster_id", "access_token", "ray_config", "node_info", "raylet_info",
-        "tune_info", "tune_availability"
-    ]
+    __schema__ = {
+        "cluster_id": (True, None, str),
+        "access_token": (True, None, str),
+        "ray_config": (True, None, tuple),
+        "node_info": (True, None, dict),
+        "raylet_info": (True, None, dict),
+        "tune_info": (True, None, dict),
+        "tune_availability": (True, None, dict)
+    }
 
 
-# TODO(sang): Add piggybacked response.
 class IngestResponse(BaseModel):
-    pass
+    __schema__ = {"succeed": (True, None, bool), "actions": (False, [], list)}
 
 
 class AuthRequest(BaseModel):
-    __slots__ = ["cluster_id"]
+    __schema__ = {"cluster_id": (True, None, str)}
 
 
 class AuthResponse(BaseModel):
-    __slots__ = ["dashboard_url", "access_token"]
+    __schema__ = {
+        "dashboard_url": (True, None, str),
+        "access_token": (True, None, str)
+    }
+
+
+"""
```

Comment:
No need to add section here if the file is 100 lines

Reply:
Will remove